ai2thor/controller.py

```diff
@@ -410,6 +411,22 @@ def __init__(
         self.add_depth_noise = add_depth_noise
         self.include_private_scenes = include_private_scenes
         self.x_display = None
+        self.gpu_device = gpu_device
+        cuda_visible_devices = list(
+            map(int, filter(lambda y: y.isdigit(),
+            map(lambda x: x.strip(),
+                os.environ.get('CUDA_VISIBLE_DEVICES', '').split(',')))))
+
+        if self.gpu_device:
+            if (type(self.gpu_device) is not int or self.gpu_device < 0):
+                raise ValueError("Invalid gpu_device: '%s'. gpu_device must be >= 0" % self.gpu_device)
+            elif cuda_visible_devices:
+                if self.gpu_device >= len(cuda_visible_devices):
+                    raise ValueError("Invalid gpu_device: '%s'. gpu_device must less than number of CUDA_VISIBLE_DEVICES: %s" % (self.gpu_device, cuda_visible_devices))
+                else:
+                    self.gpu_device = cuda_visible_devices[self.gpu_device]
```

Comment:
Oh I see, this seems a bit weird to me, if I specify `self.gpu_device == 3` then I'd expect gpu 3 to be used and not GPU 7 if I had specified `cuda_visible_devices=[3,4,5,7]`. That said, if this is a common convention (indexing into `cuda_visible_devices` rather than using absolute indices) then I'm happy with that.

Reply:
Yes, this was what the preference was from testing with users at CMU.